File: mesop/components/date_range_picker/date_range_picker.ts
import {
  MatDatepickerModule,
  MatDatepickerInputEvent,
  DateRange,
} from '@angular/material/datepicker';
import {MatFormFieldModule} from '@angular/material/form-field';
import {MatInputModule} from '@angular/material/input';
import {Component, Input} from '@angular/core';
import {
  FormControl,
  FormGroup,
  FormsModule,
  ReactiveFormsModule,
} from '@angular/forms';
import {
  DateRangePickerEvent,
  UserEvent,
  Key,
  Type,
  Style,
} from 'mesop/mesop/protos/ui_jspb_proto_pb/mesop/protos/ui_pb';
import {DateRangePickerType} from 'mesop/mesop/components/date_range_picker/date_range_picker_jspb_proto_pb/mesop/components/date_range_picker/date_range_picker_pb';
import {CommonModule} from '@angular/common';
import {Channel} from '../../web/src/services/channel';
import {formatStyle} from '../../web/src/utils/styles';
import {provideNativeDateAdapter} from '@angular/material/core';
import {debounceTime} from 'rxjs/operators';
import {Subject} from 'rxjs';

@Component({
  selector: 'mesop-date-range-picker',
  templateUrl: 'date_range_picker.ng.html',
  standalone: true,
  providers: [provideNativeDateAdapter()],
  imports: [
    MatDatepickerModule,
    MatFormFieldModule,
    MatInputModule,
    CommonModule,
    FormsModule,
    ReactiveFormsModule,
  ],
})
export class DateRangePickerComponent {
  @Input({required: true}) type!: Type;
  @Input() key!: Key;
  @Input() style!: Style;
  private _config!: DateRangePickerType;

  readonly dateRange = new FormGroup({
    start: new FormControl(),
    end: new FormControl(),
  });
  private initialStartDate = '';
  private initialEndDate = '';
  private dateRangeChangeSubject = new Subject<
    MatDatepickerInputEvent<string | undefined, DateRange<string | undefined>>
  >();

  constructor(private readonly channel: Channel) {
    this.dateRangeChangeSubject
      .pipe(debounceTime(30))
      .subscribe((event) => this.onChangeDebounced(event));
  }

  ngOnChanges() {
    this._config = DateRangePickerType.deserializeBinary(
      this.type.getValue() as unknown as Uint8Array,
    );

    const startDateString = this._config.getStartDate() || '';
    const endDateString = this._config.getEndDate() || '';

    let dateInputUpdated = false;
    if (this.initialStartDate !== startDateString) {
      this.initialStartDate = startDateString;
      dateInputUpdated = true;
    }
    if (this.initialEndDate !== endDateString) {
      this.initialEndDate = endDateString;
      this.initialEndDate = endDateString;
      dateInputUpdated = true;
    }

    if (dateInputUpdated) {
      this.dateRange.setValue({
        start: this.makeDate(startDateString),
        end: this.makeDate(endDateString),
      });
    }
  }

  onChange(
    event: MatDatepickerInputEvent<
      string | undefined,
      DateRange<string | undefined>
    >,
  ): void {
    this.dateRangeChangeSubject.next(event);
  }

  // We debounce the date range change event to avoid an issue where two change events
  // are fired consecutively when resetting the date range.
  //
  // If we send both events, this can cause issues if the value parameters are set on
  // the component. What will happen is the end date will not get reset, which is why
  // we only want to send the second event.
  //
  // The other reason we need debounce is to handle the case where the user edits the
  // start date directly. On blur, only one event will fire. And we need to send this
  // event if the end date is also set.
  onChangeDebounced(
    event: MatDatepickerInputEvent<
      string | undefined,
      DateRange<string | undefined>
    >,
  ): void {
    if (this.dateRange.value.start && this.dateRange.value.end) {
      const userEvent = new UserEvent();
      userEvent.setHandlerId(this.config().getOnChangeHandlerId()!);
      const dateRangeEvent = new DateRangePickerEvent();
      dateRangeEvent.setStartDate(
        this.makeDateString(this.dateRange.value.start),
      );
      dateRangeEvent.setEndDate(this.makeDateString(this.dateRange.value.end));
      userEvent.setDateRangePickerEvent(dateRangeEvent);
      userEvent.setKey(this.key);
      this.channel.dispatch(userEvent);
    }
  }

  private makeDate(dateInput: string): Date | null {
    if (dateInput) {
      const [year, month, date] = dateInput.split('-').map((e) => parseInt(e));
      return new Date(year, month - 1, date);
    }
    return null;
  }

  private makeDateString(date: Date | null): string {
    return date
      ? `${date.getFullYear()}-${date.getMonth() + 1}-${date.getDate()}`
      : '';
  }

  config(): DateRangePickerType {
    return this._config;
  }

  getStyle(): string {
    return formatStyle(this.style);
  }

  getColor(): 'primary' | 'accent' | 'warn' {
    return this.config().getColor() as 'primary' | 'accent' | 'warn';
  }

  getFloatLabel(): 'always' | 'auto' {
    return this.config().getFloatLabel() as 'always' | 'auto';
  }

  getAppearance(): 'fill' | 'outline' {
    return this.config().getAppearance() as 'fill' | 'outline';
  }

  getSubscriptSizing(): 'fixed' | 'dynamic' {
    return this.config().getSubscriptSizing() as 'fixed' | 'dynamic';
  }
}

File: mesop/components/date_range_picker/date_range_picker.py
from dataclasses import dataclass
from datetime import date, datetime
from typing import Any, Callable, Literal

import mesop.components.date_range_picker.date_range_picker_pb2 as date_range_picker_pb
from mesop.component_helpers import (
  Style,
  insert_component,
  register_event_handler,
  register_event_mapper,
  register_native_component,
)
from mesop.events import MesopEvent

_DATE_FORMAT = "%Y-%m-%d"


@dataclass(kw_only=True)
class DateRangePickerChangeEvent(MesopEvent):
  """Represents a date range picker change event.

  This event will only fire if start and end dates are valid dates.

  Attributes:
      start_date: Start date
      end_date: End date
      key (str): key of the component that emitted this event.
  """

  start_date: date
  end_date: date


register_event_mapper(
  DateRangePickerChangeEvent,
  lambda userEvent, key: DateRangePickerChangeEvent(
    start_date=_try_make_date(userEvent.date_range_picker_event.start_date),
    end_date=_try_make_date(userEvent.date_range_picker_event.end_date),
    key=key.key,
  ),
)


@register_native_component
def date_range_picker(
  *,
  label: str = "",
  on_change: Callable[[DateRangePickerChangeEvent], Any] | None = None,
  appearance: Literal["fill", "outline"] = "fill",
  style: Style | None = None,
  disabled: bool = False,
  placeholder_start_date: str = "",
  placeholder_end_date: str = "",
  required: bool = False,
  start_date: date | None = None,
  end_date: date | None = None,
  readonly: bool = False,
  hide_required_marker: bool = False,
  color: Literal["primary", "accent", "warn"] = "primary",
  float_label: Literal["always", "auto"] = "auto",
  subscript_sizing: Literal["fixed", "dynamic"] = "fixed",
  hint_label: str = "",
  key: str | None = None,
):
  """Creates a date range picker component.

  Args:
    label: Label for date range picker input.
    on_change: Fires when valid date values for both start/end have been specified through Calendar date selection or user input blur.
    appearance: The form field appearance style.
    style: Style for date range picker input.
    disabled: Whether it's disabled.
    placeholder_start_date: Start date placeholder value.
    placeholder_end_date: End date placeholder value.
    required: Whether it's required.
    start_date: Start date initial value.
    end_date: End date initial value.
    readonly: Whether the element is readonly.
    hide_required_marker: Whether the required marker should be hidden.
    color: The color palette for the form field.
    float_label: Whether the label should always float or float as the user types.
    subscript_sizing: Whether the form field should reserve space for one line of hint/error text (default) or to have the spacing grow from 0px as needed based on the size of the hint/error content. Note that when using dynamic sizing, layout shifts will occur when hint/error text changes.
    hint_label: Text for the form field hint.
    key: The component [key](../components/index.md#component-key).
  """

  insert_component(
    key=key,
    type_name="date_range_picker",
    proto=date_range_picker_pb.DateRangePickerType(
      start_date=_try_make_date_str(start_date),
      end_date=_try_make_date_str(end_date),
      placeholder_start_date=placeholder_start_date,
      placeholder_end_date=placeholder_end_date,
      disabled=disabled,
      required=required,
      readonly=readonly,
      hide_required_marker=hide_required_marker,
      color=color,
      float_label=float_label,
      appearance=appearance,
      subscript_sizing=subscript_sizing,
      hint_label=hint_label,
      label=label,
      on_change_handler_id=register_event_handler(
        on_change, event=DateRangePickerChangeEvent
      )
      if on_change
      else "",
    ),
    style=style,
  )


def _try_make_date_str(date_input: date | None) -> str:
  if date_input:
    return date_input.isoformat()
  return ""


def _try_make_date(date_string: str) -> date:
  return datetime.strptime(date_string, _DATE_FORMAT).date()
